### PR TITLE
Online SLAM Example with Realsense sensor support, bagfile support, and saved dataset support.

### PR DESCRIPTION
**Pull request by reyanshsolis**
# Use RealSense Sensor for Live Online Reconstruction
To use this example:
```bash
# build open3d with Librealsense support (and CUDA if available)
cd build
cmake  -DBUILD_CUDA_MODULE=ON -DBUILD_LIBREALSENSE=ON .. 
# make the example
make OnlineSLAMRealSense OnlineSLAMRGBD
```
- Download [RealSense config file `rs_default_config.json`](https://github.com/isl-org/open3d_downloads/releases/download/20220301-data/rs_default_config.json) and modify the same according to your sensor. 
- Connect the ReaSense sensor and run the example
```bash
cd build
./bin/examples/OnlineSLAMRealSense --config /path/rs_default_config.json --align --device "CUDA:0"
```
# Use RealSense Bag File.
```bash
./bin/examples/OnlineSLAMRealSense --config /path/rs_default_config.json --use_bag_file /path/JackJackL515Bag.bag --align --device "CUDA:0"
```
# Use RGBD Dataset for Online Reconstruction
```bash
./bin/examples/OnlineSLAMRGBD /path//LoungeRGBDImages --align --device "CUDA:0"
```

This change is [Reviewable](https://reviewable.io/reviews/isl-org/open3d/4967)

### CODE REVIEW

**Comment by yxlao** [comment]
1. We should avoid duplicated code. 
    - Extact `XXXGUI.h` and `XXXGUI.cpp` from `DenseSLAMGUI.cpp`.
    - `DenseSLAMGUI.cpp` and `RealSenseDenseSLAMGUI.cpp` should use `XXXGUI.h` and `XXXGUI.cpp`.
    - CMake also needs to be updated.
2. Have you test that with the realsense sensor?

**Review by theNded — CHANGES_REQUESTED** [review]
> Reviewed 2 of 2 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/isl-org/open3d/4967)* status: all files reviewed, 2 unresolved discussions (waiting on @reyanshsolis)

___
*[examples/cpp/RealSenseDenseSLAMGUI.cpp, line 198 at r1](https://reviewable.io/reviews/isl-org/open3d/4967#-MzrMMjy0wSOgr24yxL7:-MzrMMjy0wSOgr24yxL8:bo3bcxp) ([raw file](https://github.com/isl-org/open3d/blob/c53f0556f4aec9674fe4c0695ac91f253f3bf377/examples/cpp/RealSenseDenseSLAMGUI.cpp#L198)):*
> ```cpp
>         /// Property panels
>         fixed_props_ = std::make_shared<PropertyPanel>(spacing, left_margin);
>         fixed_props_->AddIntSlider("Depth scale", &prop_values_.depth_scale,
> ```

RealSense should have built-in scale factors and it is always confusing users. Can we make a look-up table according to the types (check with the previous realsense team maybe?), and fix it according to the type on launch?
___
*[examples/cpp/RealSenseDenseSLAMGUI.cpp, line 465 at r1](https://reviewable.io/reviews/isl-org/open3d/4967#-MzrMrVADNX_XSXrcqIz:-MzrMrVADNX_XSXrcqJ-:bvajbei) ([raw file](https://github.com/isl-org/open3d/blob/c53f0556f4aec9674fe4c0695ac91f253f3bf377/examples/cpp/RealSenseDenseSLAMGUI.cpp#L465)):*
> ```cpp
> 
>         // Only set at initialization
>         float depth_scale = prop_values_.depth_scale;
> ```

Same as above, set this according to the rs profile.

**Review by reyanshsolis — COMMENTED** [review]
1. Done.
2. Yes.

> *[Reviewable](https://reviewable.io/reviews/isl-org/open3d/4967#-:-Mzzap8W4oYGGZ-wBb65:bwi7iwf)* status: 0 of 9 files reviewed, 2 unresolved discussions (waiting on @theNded)

___
*[examples/cpp/RealSenseDenseSLAMGUI.cpp, line 198 at r1](https://reviewable.io/reviews/isl-org/open3d/4967#-MzrMMjy0wSOgr24yxL7:-Mzzavi-9Ls6J8Hwdevd:b-eupni3) ([raw file](https://github.com/isl-org/open3d/blob/c53f0556f4aec9674fe4c0695ac91f253f3bf377/examples/cpp/RealSenseDenseSLAMGUI.cpp#L198)):*
<details><summary><i>Previously, theNded (Wei Dong) wrote…</i></summary><blockquote>

RealSense should have built-in scale factors and it is always confusing users. Can we make a look-up table according to the types (check with the previous realsense team maybe?), and fix it according to the type on launch?
</blockquote></details>

Done. Provided a `default_param` lookup table param, which is used to pass specific default values from sensors.
___
*[examples/cpp/RealSenseDenseSLAMGUI.cpp, line 465 at r1](https://reviewable.io/reviews/isl-org/open3d/4967#-MzrMrVADNX_XSXrcqIz:-MzzbDToDpZOTcQdb_mg:b-896fix) ([raw file](https://github.com/isl-org/open3d/blob/c53f0556f4aec9674fe4c0695ac91f253f3bf377/examples/cpp/RealSenseDenseSLAMGUI.cpp#L465)):*
<details><summary><i>Previously, theNded (Wei Dong) wrote…</i></summary><blockquote>

Same as above, set this according to the rs profile.
</blockquote></details>

Done.

**Review by reyanshsolis — COMMENTED** [review]
1. Done.
2. Yes.

> *[Reviewable](https://reviewable.io/reviews/isl-org/open3d/4967#-:-Mzzap8W4oYGGZ-wBb65:bwi7iwf)* status: 0 of 9 files reviewed, 2 unresolved discussions (waiting on @theNded)

___
*[examples/cpp/RealSenseDenseSLAMGUI.cpp, line 198 at r1](https://reviewable.io/reviews/isl-org/open3d/4967#-MzrMMjy0wSOgr24yxL7:-Mzzavi-9Ls6J8Hwdevd:b-eupni3) ([raw file](https://github.com/isl-org/open3d/blob/c53f0556f4aec9674fe4c0695ac91f253f3bf377/examples/cpp/RealSenseDenseSLAMGUI.cpp#L198)):*
<details><summary><i>Previously, theNded (Wei Dong) wrote…</i></summary><blockquote>

RealSense should have built-in scale factors and it is always confusing users. Can we make a look-up table according to the types (check with the previous realsense team maybe?), and fix it according to the type on launch?
</blockquote></details>

Done. Provided a `default_param` lookup table param, which is used to pass specific default values from sensors.
___
*[examples/cpp/RealSenseDenseSLAMGUI.cpp, line 465 at r1](https://reviewable.io/reviews/isl-org/open3d/4967#-MzrMrVADNX_XSXrcqIz:-MzzbDToDpZOTcQdb_mg:b-896fix) ([raw file](https://github.com/isl-org/open3d/blob/c53f0556f4aec9674fe4c0695ac91f253f3bf377/examples/cpp/RealSenseDenseSLAMGUI.cpp#L465)):*
<details><summary><i>Previously, theNded (Wei Dong) wrote…</i></summary><blockquote>

Same as above, set this according to the rs profile.
</blockquote></details>

Done.

**Review by reyanshsolis — COMMENTED** [review]
1. Done.
2. Yes.

> *[Reviewable](https://reviewable.io/reviews/isl-org/open3d/4967#-:-Mzzap8W4oYGGZ-wBb65:bwi7iwf)* status: 0 of 9 files reviewed, 2 unresolved discussions (waiting on @theNded)

___
*[examples/cpp/RealSenseDenseSLAMGUI.cpp, line 198 at r1](https://reviewable.io/reviews/isl-org/open3d/4967#-MzrMMjy0wSOgr24yxL7:-Mzzavi-9Ls6J8Hwdevd:b-eupni3) ([raw file](https://github.com/isl-org/open3d/blob/c53f0556f4aec9674fe4c0695ac91f253f3bf377/examples/cpp/RealSenseDenseSLAMGUI.cpp#L198)):*
<details><summary><i>Previously, theNded (Wei Dong) wrote…</i></summary><blockquote>

RealSense should have built-in scale factors and it is always confusing users. Can we make a look-up table according to the types (check with the previous realsense team maybe?), and fix it according to the type on launch?
</blockquote></details>

Done. Provided a `default_param` lookup table param, which is used to pass specific default values from sensors.
___
*[examples/cpp/RealSenseDenseSLAMGUI.cpp, line 465 at r1](https://reviewable.io/reviews/isl-org/open3d/4967#-MzrMrVADNX_XSXrcqIz:-MzzbDToDpZOTcQdb_mg:b-896fix) ([raw file](https://github.com/isl-org/open3d/blob/c53f0556f4aec9674fe4c0695ac91f253f3bf377/examples/cpp/RealSenseDenseSLAMGUI.cpp#L465)):*
<details><summary><i>Previously, theNded (Wei Dong) wrote…</i></summary><blockquote>

Same as above, set this according to the rs profile.
</blockquote></details>

Done.

**Review by yxlao — CHANGES_REQUESTED** [review]
> Reviewed 5 of 9 files at r2, 3 of 6 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/isl-org/open3d/4967)* status: 8 of 11 files reviewed, 6 unresolved discussions (waiting on @reyanshsolis, @theNded, and @yxlao)

___
*[cpp/open3d/camera/PinholeCameraIntrinsic.h, line 58 at r3](https://reviewable.io/reviews/isl-org/open3d/4967#-N-8H7152f7a1aE0zrXO:-N-8H7152f7a1aE0zrXP:b-qu244g) ([raw file](https://github.com/isl-org/open3d/blob/1805e277903de4ebb904d7a176218d5704e9514d/cpp/open3d/camera/PinholeCameraIntrinsic.h#L58)):*
> ```objc
>     /// \param height height of the image. (Default: -1).
>     /// \param intrinsic_matrix 3x3 intrinsic matrix. (Default: Identity).
>     PinholeCameraIntrinsic(int width = -1,
> ```

These defaults can be confusing. `PinholeCameraIntrinsic(int width, int height, double fx, double fy, double cx, double cy);` does not have defaults. 

Recommend:
1. Keep original `PinholeCameraIntrinsic();`
2. Add `PinholeCameraIntrinsic(int, int, Eigen::Matrix3d)`, with no default values
3. `const Eigen::Matrix3d &` with `&`.
___
*[examples/cpp/OnlineSLAMUtil.h, line 34 at r3](https://reviewable.io/reviews/isl-org/open3d/4967#-N-8I_0H9BlptZbdzlpK:-N-8I_0H9BlptZbdzlpL:b-rhb2ir) ([raw file](https://github.com/isl-org/open3d/blob/1805e277903de4ebb904d7a176218d5704e9514d/examples/cpp/OnlineSLAMUtil.h#L34)):*
> ```objc
> 
> // Tanglo colorscheme (see https://en.wikipedia.org/wiki/Tango_Desktop_Project)
> static const Eigen::Vector3d kTangoOrange(0.961, 0.475, 0.000);
> ```

These can be inside the namesapce.
___
*[examples/cpp/OnlineSLAMUtil.h, line 38 at r3](https://reviewable.io/reviews/isl-org/open3d/4967#-N-8Ie-50jJY3yKPR68g:-N-8Ie-50jJY3yKPR68h:b3zf9rb) ([raw file](https://github.com/isl-org/open3d/blob/1805e277903de4ebb904d7a176218d5704e9514d/examples/cpp/OnlineSLAMUtil.h#L38)):*
> ```objc
> 
> namespace open3d {
> namespace online_slam {
> ```

Does it make sense to call this `open3d::example` or `open3d::example::online_slam`?
___
*[examples/cpp/OnlineSLAMUtil.h, line 159 at r3](https://reviewable.io/reviews/isl-org/open3d/4967#-N-8IpoE3T_hT5YSBRj_:-N-8IpoF3_4NT1XVObgo:b5b4o0y) ([raw file](https://github.com/isl-org/open3d/blob/1805e277903de4ebb904d7a176218d5704e9514d/examples/cpp/OnlineSLAMUtil.h#L159)):*
> ```objc
> };
> 
> //------------------------------------------------------------------------------
> ```

separator not necessary

**Review by reyanshsolis — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/isl-org/open3d/4967)* status: 5 of 11 files reviewed, 6 unresolved discussions (waiting on @reyanshsolis, @theNded, and @yxlao)

___
*[cpp/open3d/camera/PinholeCameraIntrinsic.h, line 58 at r3](https://reviewable.io/reviews/isl-org/open3d/4967#-N-8H7152f7a1aE0zrXO:-N-E4yUm-XicxtzIjOD4:b-896fix) ([raw file](https://github.com/isl-org/open3d/blob/1805e277903de4ebb904d7a176218d5704e9514d/cpp/open3d/camera/PinholeCameraIntrinsic.h#L58)):*
<details><summary><i>Previously, yxlao (Yixing Lao) wrote…</i></summary><blockquote>

These defaults can be confusing. `PinholeCameraIntrinsic(int width, int height, double fx, double fy, double cx, double cy);` does not have defaults. 

Recommend:
1. Keep original `PinholeCameraIntrinsic();`
2. Add `PinholeCameraIntrinsic(int, int, Eigen::Matrix3d)`, with no default values
3. `const Eigen::Matrix3d &` with `&`.

</blockquote></details>

Done.
___
*[examples/cpp/OnlineSLAMUtil.h, line 38 at r3](https://reviewable.io/reviews/isl-org/open3d/4967#-N-8Ie-50jJY3yKPR68g:-N-E55Yh5ToP2S5zc8hn:b-6ovvbl) ([raw file](https://github.com/isl-org/open3d/blob/1805e277903de4ebb904d7a176218d5704e9514d/examples/cpp/OnlineSLAMUtil.h#L38)):*
<details><summary><i>Previously, yxlao (Yixing Lao) wrote…</i></summary><blockquote>

Does it make sense to call this `open3d::example` or `open3d::example::online_slam`?
</blockquote></details>

Done. using `open3d::examples::online_slam` (the dir struct is `examples`).
___
*[examples/cpp/OnlineSLAMUtil.h, line 159 at r3](https://reviewable.io/reviews/isl-org/open3d/4967#-N-8IpoE3T_hT5YSBRj_:-N-E5EvU2G8i963CVIsw:b-896fix) ([raw file](https://github.com/isl-org/open3d/blob/1805e277903de4ebb904d7a176218d5704e9514d/examples/cpp/OnlineSLAMUtil.h#L159)):*
<details><summary><i>Previously, yxlao (Yixing Lao) wrote…</i></summary><blockquote>

separator not necessary
</blockquote></details>

Done.

**Review by reyanshsolis — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/isl-org/open3d/4967)* status: 5 of 11 files reviewed, 6 unresolved discussions (waiting on @theNded and @yxlao)

___
*[examples/cpp/OnlineSLAMUtil.h, line 34 at r3](https://reviewable.io/reviews/isl-org/open3d/4967#-N-8I_0H9BlptZbdzlpK:-N-E6gK93K6rMO9tOQh7:b-896fix) ([raw file](https://github.com/isl-org/open3d/blob/1805e277903de4ebb904d7a176218d5704e9514d/examples/cpp/OnlineSLAMUtil.h#L34)):*
<details><summary><i>Previously, yxlao (Yixing Lao) wrote…</i></summary><blockquote>

These can be inside the namesapce.
</blockquote></details>

Done.

File: examples/cpp/OnlineSLAMRGBD.cpp [hunk 1]
```diff
@@ -0,0 +1,184 @@
...
+                "mismatch!",
+                dataset_path);
+    }
+
+    std::sort(rgb_files.begin(), rgb_files.end());
+    std::sort(depth_files.begin(), depth_files.end());
+    return std::make_pair(rgb_files, depth_files);
+}
+
+//------------------------------------------------------------------------------
```
Comment (yxlao) [inline examples/cpp/OnlineSLAMRGBD.cpp:81]: remove comments

File: examples/cpp/OnlineSLAMRealSense.cpp [hunk 2]
```diff
@@ -0,0 +1,172 @@
...
+// ----------------------------------------------------------------------------
+
+#include <chrono>
+#include <memory>
+#include <string>
+
+#include "OnlineSLAMUtil.h"
+#include "open3d/Open3D.h"
+
+//------------------------------------------------------------------------------
```
Comment (yxlao) [inline examples/cpp/OnlineSLAMRealSense.cpp:34]: remove comments

**Review by yxlao — APPROVED** [review]

**Review by reyanshsolis — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/isl-org/open3d/4967)* status: 5 of 11 files reviewed, 8 unresolved discussions (waiting on @theNded and @yxlao)

___
*[examples/cpp/OnlineSLAMRealSense.cpp, line 34 at r4](https://reviewable.io/reviews/isl-org/open3d/4967#-N-7OSOzX_E_wIEKUEP9-r4-34:-N-HHUKI9PUK_48_5Ncl:b-896fix) ([raw file](https://github.com/isl-org/open3d/blob/2e6fa2a9cc3a66501cb94742c41d663a7fa48791/examples/cpp/OnlineSLAMRealSense.cpp#L34)):*
<details><summary><i>Previously, yxlao (Yixing Lao) wrote…</i></summary><blockquote>

remove comments
</blockquote></details>

Done.
___
*[examples/cpp/OnlineSLAMRGBD.cpp, line 81 at r4](https://reviewable.io/reviews/isl-org/open3d/4967#-N-7OSOzX_E_wIEKUEP8-r4-81:-N-HHTuy40jvYnVTF3n8:b-896fix) ([raw file](https://github.com/isl-org/open3d/blob/2e6fa2a9cc3a66501cb94742c41d663a7fa48791/examples/cpp/OnlineSLAMRGBD.cpp#L81)):*
<details><summary><i>Previously, yxlao (Yixing Lao) wrote…</i></summary><blockquote>

remove comments
</blockquote></details>

Done.

**Pull request merged**